### ISSUE TITLE
Accessing state after dispatching an action inside getInitialProps

### ISSUE THREAD

**Issue by ussuritiger** [bmealhouse/next-redux-saga#20]
I need to set an HTTP status inside `getInitialProps` based on data fetched by dispatching an action. So basically I need to execute the following actions synchronously:
1) Dispatch an action to fetch data
2) Based on state (fetched data) I need to adjust HTTP response status code

Since dispatching an action that will be picked up by saga is done asynchronously, I've ended up with the following:
1) Created a method `stopSagaTask` while configuring store (does the same as `next-redux-saga` does to figure out when sagas are completed):
```javascript
  store.stopSagaTask = async () => {
    store.dispatch(END)

    await store.sagaTask.done
  }
```
2) Adjusted `getInitialProps` to stop sagas after dispatching an action so we can access state when sagas are completed but inside the `getInitialProps` method where we still have access to `request` and `response`:
```javascript
Component.getInitialProps = async ctx => {
  const { store, query } = ctx

  store.dispatch(fetchSomethingById(query.id))

  await store.stopSagaTask()

  console.log(store.getState()) // we can modify response HTTP status based on fetched data
}
```
So this works but I am wondering if this is the most optimal way.

**Comment by cJayyy** [comment]
I am stuck in the same problem. 
Looking forward to see any better solution. 

@ussuritiger thanks for pointing out this workaround.

**Comment by bmealhouse** [comment]
Sorry for the late response @ussuritiger. I don't use redux-saga anymore, so I'm probably not the best person to answer your question. I'll leave this issue open and let others from the community chime in if they have any advice.

**Comment by bbortt** [comment]
@ussuritiger I know this is a very old thread, but here are my thoughts.
First of all, using redux-saga in synchronous mode will no longer be possible starting from `redux-saga: 1.0.0`. A closed stdChannel (`store.dispatch(END)`) cannot be reopened.
Second, what you try there is very much against the idea of redux-saga. It is designed to handle side effects *asynchronously*. The most common way to handle the response is in the saga its self:
```
const response = yield call(axios.get, `some-url`)
yield call(handleResponse, response.data.)
```
I think you may close this issue as it is way out of date.

**Comment by cJayyy** [comment]
@bbortt But how would you initiate async props if not inside `getInitialProps`?

**Comment by bbortt** [comment]
@cJayyy I think `getInitialProps` is still the right place to trigger sagas. This is what you're asking for if i understand you right?
Sagas will then load the data and dispatch another event, for example `FETCH_SUCCESS`. This event then reduces an updated state. And thanks to how react works this will update the data in your component.
Subscribe to the store if you need the data right away.
```
if (!request.loading) {
  store.dispatch(event(properties))
}

const unsubscribe = store.subscribe(() => {
    const updatedRequest = store.getState().request

    // Do whatever you want here

    unsubscribe()
})
```

**Comment by cJayyy** [comment]
@bbortt That's right, but my question was rather about how can we load async data before render (and in this case - on a server-side)?
The best solution is currently to use `getInitialProps` but since we dispatch action (which obviously doesn't block the flow of `getInitialProps`) we can't track the state of async function therefore component starts rendering without any chance to know if the async is already done.

**Comment by dorian-marchal** [comment]
@cJayyy: Expanding on @bbortt example:

```js
// ...somewhere in your `makeStore` function:
sagaMiddleware.run(function*() {
  yield all([
    takeLatest('PAGE_LOAD_REQUEST', function* loadDeps(action) {
      let deps = null;
      try {
        deps = yield call(yourFetchFunction, action.payload);
      } catch (err) {
        yield put({ type: 'PAGE_LOAD_ERROR', error: true });
        return;
      }
      yield put({ type: 'PAGE_LOAD_SUCCESS', payload: deps });
    }),
  ]);
});
```
```js
// ... in your reducer:
const initialState = { hasLoadingFailed: false, isPageLoading: false };
function reducer(state = initialState, action) {
  switch (action.type) {
    case 'PAGE_LOAD_REQUEST':
      return {
        ...state,
        hasLoadingFailed: false,
        isPageLoading: true,
      };

    case 'PAGE_LOAD_SUCCESS':
    case `PAGE_LOAD_ERROR`:
      return {
        ...state,
        hasLoadingFailed: Boolean(action.error),
        isPageLoading: false,
      };
  }

  return state;
}
```
```js
// ...in `getInitialProps`:
static async getInitialProps({ ctx }) {
  const { store } = ctx;

  store.dispatch({ type: 'PAGE_LOAD_REQUEST', payload: 'whatever you want, probably the pathname' });

  // Wait for your dependencies to be resolved.
  await new Promise((resolve) => {
    const unsubscribe = store.subscribe(() => {
      const state = store.getState();
      if (!state.isPageLoading) {
        unsubscribe();
        resolve();
      }
    });
  });
}
```

This way, `getInitialProps` only resolves when your data are fetched.

**Comment by ussuritiger** [comment]
Thanks guys for sharing. Solution described by @bbortt and @dorian-marchal looks optimal and indeed adhere to the nature of redux-saga. I've tried it and it works well for me.

I will close this issue however it is open for any other ideas.

**Comment by bbortt** [comment]
thank you @dorian-marchal I've probably kept my explanation a little to short :)
glad we could help @ussuritiger

**Comment by cJayyy** [comment]
Thanks a lot!
This solution looks pretty good to me. :)  👍

**Comment by emirotin** [comment]
While I understand this thread I'd like to clarify. This thread demonstrates how to await for data load in getInitialProps should I want to do something with the store there (like for example we make some redirects).
But if I _don't_ do this `subscribe` ting this library still guarantees that _server-side_ all async actions/sagas are done before `render` is called. Or not?

**Comment by bbortt** [comment]
@emirotin Yes, it does.

Given this source code
```
// Stop saga on the server
      if (isServer) {
        store.dispatch(END)
        await store.sagaTask.toPromise()
}
```
And the corresponding test `next-redux-saga/test/server.test.js`
```
test('Wrapped component awaits asynchronous getInitialProps', async () => {
  const WrappedComponent = withRedux(configureStore)(
    withReduxSaga(AsyncGetInitialProps),
  )

  const props = await getInitialProps(WrappedComponent)
  expect(props.isServer).toBeTruthy()
  expect(props.initialState).toEqual({
    syncReduxProp: SYNC_REDUX_PROP_TEXT,
    asyncReduxSagaProp: ASYNC_REDUX_SAGA_PROP_TEXT,
  })
  expect(props.initialProps).toEqual({staticProp: STATIC_PROP_TEXT})

  createSnapshot(WrappedComponent, props)
})
```

**Comment by emirotin** [comment]
Yeah good, that's what I understood. Just wanted to make sure as after migrating to sagas I'm seeing strange issues, but with client-side navigation. Would open a new issue but didn't isolate it yet in a simplified demo app.

**Comment by bbortt** [comment]
@emirotin no problem. let me know if I can help.

**Comment by emirotin** [comment]
Thanks @bbortt 
Until I can create a repro maybe the symptoms will tell you something.
So SSR always works fine. But when after that I do client-side navigation often the page becomes empty. Sometimes in the network inspector I see the target page's bundle returned with realistic size but the browser tells it cannot read the response

This happens more or less consistently but not on every navigation, and it has never happened before I started the migration to sagas.

**Comment by bbortt** [comment]
@emirotin this doesn't look like anything to me ;) please open a new issue as soon as you've any news 👍

**Comment by ioness** [comment]
When we want to await for a saga to finish, we execute it directly, without dispatching an action..

export function executeSaga (saga, ...args) {
  return sagaMiddleware.run(saga, ...args).done
}

That way sagas are responsible for all side effects in the project.

**Comment by dimisus** [comment]
> This way, getInitialProps only resolves when your data are fetched. @dorian-marchal 

Will this work on the clients side ?

My problem: 

async getInitialProps only wait for my sagas to end if I call an async action inside the method - on the server side i do dispatchEND and await sagas.toPromise(), which works great on first page load.

On the clients side though I want to use getInitialProps to hydrate my store and return pageProps. This way I would render only once, but I cannot dispatchEND and await sagas.toPromise() on the client. All the sagas are unusable afterwards.

**Comment by bbortt** [comment]
@dimisus no, this is not possible. `saga` is designed to be used asynchronous. store updates will be delivered continuously.

**Comment by iamdevlinph** [comment]
> @cJayyy: Expanding on @bbortt example:
> 
> ```js
> // ...in `getInitialProps`:
>   // Wait for your dependencies to be resolved.
>   await new Promise((resolve) => {
>     const unsubscribe = store.subscribe(() => {
>       const state = store.getState();
>       if (!state.isPageLoading) {
>         unsubscribe();
>         resolve();
>       }
>     });
>   });
> }
> ```
> 

Adding the code above seems to make the page not load. It's just keeps on loading without any error.

Is there any other way to achieve this?

**Comment by emirotin** [comment]
The fact that your page does not finish loading is likely an indication something is wrong with your saga, either you're not receiving the data (remote endpoint issue) or you're not properly clearing the `isPageLoading` flag (should it actually be on the root state object and not nested under some _combine_d reducer key?)

**Comment by iamdevlinph** [comment]
I have narrowed down the code to just:
```js
// Wait for your dependencies to be resolved.
await new Promise(resolve => {
  const unsubscribe = store.subscribe(() => {
    unsubscribe();
    resolve();
  });
});
```
But the issue is still there.

The saga I'm calling is connecting to a firebase though. I'll try to use a basic flow for now to test.

**Comment by iamdevlinph** [comment]
Oh wait. I don't think this is related to sagas.

I have a dispatch just above the `new Promise` code. And the issue only appears when the `store.subscribe()` code is present.

**Comment by emirotin** [comment]
You can try a helper I've created for this:

```
export const ensureState = async ({ condition, store }) => {
    const state = store.getState();
    if (condition(state)) {
        return state;
    }

    return new Promise(resolve => {
        const unsubscribe = store.subscribe(() => {
            const state = store.getState();
            if (condition(state)) {
                unsubscribe();
                resolve(state);
            }
        });
    });
};
```

It's also doing the immediate check but given the single-threaded nature of JS it shouldn't be making any difference.

With your most recent example, you'd see the page blocked if the store is never updated since you dispatch the sagas.

**Comment by emirotin** [comment]
> Oh wait. I don't think this is related to sagas.
> 
> I have a dispatch just above the `new Promise` code. And the issue only appears when the `store.subscribe()` code is present.

This is the correct code structure and in my understanding, it means your sagas end up not dispatching a single action (that would trigger the subscription).

**Comment by iamdevlinph** [comment]
Let me provide more code.

The issue occurs in this code.

```js
Page.getInitialProps = async ({ ctx }) => {
  const { id } = ctx.query;
  const { store } = ctx;
  // store.dispatch(blogActions.requestPost(id));

  // Wait for your dependencies to be resolved.
  await new Promise(resolve => {
    const unsubscribe = store.subscribe(() => {
      unsubscribe();
      resolve();
    });
  });

  return { id, post: { title: 'title', content: 'test' } };
};
```

But when I remove `store.subscribe()` I don't encounter any issue.

```js
Page.getInitialProps = async ({ ctx }) => {
  const { id } = ctx.query;
  // const { store } = ctx;
  // store.dispatch(blogActions.requestPost(id));

  // Wait for your dependencies to be resolved.
  await new Promise(resolve => {
    // const unsubscribe = store.subscribe(() => {
    // unsubscribe();
    resolve();
    // });
  });

  return { id, post: { title: 'title', content: 'test' } };
};
```

I have commented out the dispatch so I don't think the saga is the problem anymore.

**Comment by emirotin** [comment]
But it _is_ the problem because now nothing is updating your redux state so your subscription never fires.

You can narrow down the issue by doing some "normal" dispatch that would update the state through the reducer immediately and keep the subscription, things should work in this case.

**Comment by iamdevlinph** [comment]
Thanks. I'll give that a try.

**Comment by iamdevlinph** [comment]
I've tried a very simple dispatch. This action will just change a value from false to true. I didn't even use sagas but the issue is still there.

Though the value is in a nested state and not in a root state.

Does the value need to be in the root for the subscribe to fire?

**Comment by iamdevlinph** [comment]
But I don't think that's the case, right?

I'll try to create a bare bones repo to replicate the issue again.

**Comment by emirotin** [comment]
No, the value you're tracking can be anywhere

I'm suspicious something is wrong with your setup because in the last couple of month I've gone through all the possible guesses about what could be wrong with any lib involved (next, next-redux-wrapper, next-redux-saga) until I've found an issue in my own code :D

**Comment by iamdevlinph** [comment]
Fixed my issue. I was not passing the `initialState` to `createStore`. 

```js
function configureStore(initialState) {
  const store = createStore(
    rootReducer,
    initialState, // <--just added this one
    bindMiddleware([sagaMiddleware])
  );

  store.runSagaTask = () => {
    store.sagaTask = sagaMiddleware.run(rootSaga);
  };

  store.runSagaTask();
  return store;
}
```

Related read [here](https://github.com/zeit/next.js/issues/1927)

**Comment by emirotin** [comment]
You also seem to be creating a singleton `sagaMiddleware` outside of this method. That will hit you as sagas won't run on page reload. I had this issue myself and spent _a lot_ of time figuring out what's going on there.

**Comment by iamdevlinph** [comment]
You're right. I'm having issue now when reloading the page. It works fine when the page is accessed from a link in the app.

I've been using that code every time I do redux.

Here's the full code for reference.
```js
import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';

import rootReducer from './reducers';
import rootSaga from './sagas';

const sagaMiddleware = createSagaMiddleware();

const bindMiddleware = middleware => {
  if (process.env.NODE_ENV !== 'production') {
    /* eslint-disable */
    const { composeWithDevTools } = require('redux-devtools-extension');
    /* eslint-enable */
    return composeWithDevTools(applyMiddleware(...middleware));
  }
  return applyMiddleware(...middleware);
};

function configureStore(initialState) {
  const store = createStore(
    rootReducer,
    initialState,
    bindMiddleware([sagaMiddleware])
  );

  store.runSagaTask = () => {
    store.sagaTask = sagaMiddleware.run(rootSaga);
  };

  store.runSagaTask();
  return store;
}

export default configureStore;
```
What changes should I do to fix this?

**Comment by emirotin** [comment]
Move `const sagaMiddleware = createSagaMiddleware();` inside of the `configureStore` method.

The middleware appears to only run twice so on page reload `sagaMiddleware.run(rootSaga)` does nothing.

Also with the newest version, you don't need

```
store.runSagaTask = () => {
  store.sagaTask = sagaMiddleware.run(rootSaga);
};

store.runSagaTask();
```

Simply do `store.sagaTask = sagaMiddleware.run(rootSaga);` instead.

**Comment by bbortt** [comment]
@emirotin is right. @iamdevlinph please consider reading the latest [README.md](https://github.com/bmealhouse/next-redux-saga/blob/master/README.md).

> [1] @bbortt this has trapped me (as mentioned above) and while the example in the README is correct maybe it makes sense to emphasize this more that a fresh saga middleware must be created each time the store is created.

@emirotin I think we've written it somewhere. Give me a minute, I'll have a look at it.

**Comment by iamdevlinph** [comment]
Yes that seems to be my failure. I didn't copy the `configureStore` in the [README.md](https://github.com/bmealhouse/next-redux-saga/blob/master/README.md). I was using my own version which I used in my older projects.

**Comment by bbortt** [comment]
No problem at all @iamdevlinph.

@emirotin I've got the comment i was looking for.. it is (kind of) stated out in the `README.md`:
> /**
    \* Recreate the stdChannel (saga middleware) with every context.
     */

Introduced in [this commit](https://github.com/bmealhouse/next-redux-saga/commit/8a7ea8a4603372a78e09bc3c548b7686abe52d61). I think that's enough :)
And just in case someone has problems - it's always a pleasure to meet our community ;)

**Comment by emirotin** [comment]
Indeed it's there still I've overlooked it or didn't understand the true meaning/importance of it 😕.
And of course, when trying to recreate the issue I've been copying your entire example and didn't spot the difference with our existing code.
